src/Library.qs

```diff
@@ -19,63 +19,59 @@
         DataSize : Int);
 
     //Internal QRam type for a single QRAM Bit (is composed to form a full memory block)
-    internal newtype SingleBitQRAM = (Lookup : ((Qubit[], Qubit) => Unit is Adj + Ctl), 
-        AddressSize : Int);
+    internal newtype SingleBitQRAM = (Lookup : ((Qubit[], Qubit[]) => Unit is Adj + Ctl));
+
+    function ImplicitQRAMOracle(DataValues : (Bool[],Bool[])[]) : QRAM {
+        mutable addressSize = 0;
+        mutable valueSize = 0;
+        for((_,(address,value)) in Enumerated(DataValues)){
+            if(Length(address) > addressSize){
+                set addressSize = Length(address);
+            }
+            if(Length(value) > valueSize){
+                set valueSize = Length(value);
+            }
+        }
 
-    function ImplicitQRAMOracle(dataValues : Bool[][]) : QRAM {
-        let onesAddresses = OnesAddressesFromData(dataValues);
-        let qrams = Mapped(SingleImplicitQRAMOracle, onesAddresses); 
+        let qrams = Mapped(SingleImplicitQRAMOracle, DataValues); 
         return Default<QRAM>()
             w/ Lookup <- ApplyImplicitQRAMOracle(qrams, _, _)
-            w/ AddressSize <- (Head(qrams))::AddressSize
-            w/ DataSize <- Length(qrams);
+            w/ AddressSize <- addressSize
+            w/ DataSize <- valueSize;
+    }
+    
+    internal function SingleImplicitQRAMOracle(DataValue : (Bool[],Bool[])) : SingleBitQRAM {
```

Comment:
It looks like here that you don't use `DataValue` directly and immediately unwrap it on 44 so you could just name the arguments directly:
```suggestion
    internal function SingleImplicitQRAMOracle(address : Bool[], value : Bool[]) : SingleBitQRAM {
```
This is the same as `SingleImplicitQRAMOracle((address : Bool[], value : Bool[]))` due to singleton tuple equivalence (which I read about a while back [here](https://docs.microsoft.com/en-us/quantum/user-guide/language/types#singleton-tuple-equivalence) ).